File: python_client/fdfs/connection.py
"""
FastDFS Connection Management

This module handles TCP connections to FastDFS servers with connection pooling,
automatic reconnection, and health checking.
"""

import socket
import threading
import time
from typing import List, Optional, Dict
from contextlib import contextmanager

from .errors import NetworkError, ConnectionTimeoutError, ClientClosedError


class Connection:
    """
    Represents a TCP connection to a FastDFS server (tracker or storage).
    
    It wraps a socket with additional metadata and thread-safe operations.
    Each connection tracks its last usage time for idle timeout management.
    """
    
    def __init__(self, sock: socket.socket, addr: str):
        """
        Initialize a connection with an established socket.
        
        Args:
            sock: Connected socket
            addr: Server address in "host:port" format
        """
        self.sock = sock
        self.addr = addr
        self.last_used = time.time()
        self.lock = threading.Lock()
    
    def send(self, data: bytes, timeout: float = 30.0) -> None:
        """
        Transmits data to the server with optional timeout.
        
        This method is thread-safe and updates the lastUsed timestamp.
        
        Args:
            data: Bytes to send (must be complete message)
            timeout: Write timeout in seconds (0 means no timeout)
            
        Raises:
            NetworkError: If write fails or incomplete
        """
        with self.lock:
            try:
                if timeout > 0:
                    self.sock.settimeout(timeout)
                
                total_sent = 0
                while total_sent < len(data):
                    sent = self.sock.send(data[total_sent:])
                    if sent == 0:
                        raise NetworkError("write", self.addr, 
                                         Exception("Socket connection broken"))
                    total_sent += sent
                
                self.last_used = time.time()
            except socket.timeout as e:
                raise NetworkError("write", self.addr, e)
            except Exception as e:
                raise NetworkError("write", self.addr, e)
    
    def receive(self, size: int, timeout: float = 30.0) -> bytes:
        """
        Reads up to 'size' bytes from the server.
        
        This method may return fewer bytes than requested.
        Use receive_full if you need exactly 'size' bytes.
        
        Args:
            size: Maximum number of bytes to read
            timeout: Read timeout in seconds
            
        Returns:
            Received data (may be less than 'size')
            
        Raises:
            NetworkError: If read fails
        """
        with self.lock:
            try:
                if timeout > 0:
                    self.sock.settimeout(timeout)
                
                data = self.sock.recv(size)
                if not data:
                    raise NetworkError("read", self.addr, 
                                     Exception("Connection closed by peer"))
                
                self.last_used = time.time()
                return data
            except socket.timeout as e:
                raise NetworkError("read", self.addr, e)
            except Exception as e:
                raise NetworkError("read", self.addr, e)
    
    def receive_full(self, size: int, timeout: float = 30.0) -> bytes:
        """
        Reads exactly 'size' bytes from the server.
        
        This method blocks until all bytes are received or an error occurs.
        The timeout applies to the entire operation, not individual reads.
        
        Args:
            size: Exact number of bytes to read
            timeout: Total timeout for the operation
            
        Returns:
            Exactly 'size' bytes
            
        Raises:
            NetworkError: If read fails before receiving all bytes
        """
        with self.lock:
            try:
                if timeout > 0:
                    self.sock.settimeout(timeout)
                
                data = b''
                while len(data) < size:
                    chunk = self.sock.recv(size - len(data))
                    if not chunk:
                        raise NetworkError("read", self.addr,
                                         Exception("Connection closed by peer"))
                    data += chunk
                
                self.last_used = time.time()
                return data
            except socket.timeout as e:
                raise NetworkError("read", self.addr, e)
            except Exception as e:
                raise NetworkError("read", self.addr, e)
    
    def close(self) -> None:
        """
        Terminates the connection and releases resources.
        
        It's safe to call close multiple times.
        """
        with self.lock:
            try:
                if self.sock:
                    self.sock.close()
            except:
                pass
    
    def is_alive(self) -> bool:
        """
        Performs a non-blocking check to determine if the connection is still valid.
        
        Returns:
            True if connection appears to be alive, False otherwise
        """
        try:
            # Try to peek at the socket without removing data
            self.sock.setblocking(False)
            data = self.sock.recv(1, socket.MSG_PEEK)
            self.sock.setblocking(True)
            return True
        except BlockingIOError:
            # No data available, but connection is alive
            self.sock.setblocking(True)
            return True
        except:
            return False
    
    def get_last_used(self) -> float:
        """
        Returns the timestamp of the last send or receive operation.
        
        Returns:
            Last usage timestamp (Unix time)
        """
        with self.lock:
            return self.last_used
    
    def get_addr(self) -> str:
        """
        Returns the server address this connection is connected to.
        
        Returns:
            Address in "host:port" format
        """
        return self.addr


class ConnectionPool:
    """
    Manages a pool of reusable connections to multiple servers.
    
    It maintains separate pools for each server address and handles:
        - Connection reuse to minimize overhead
        - Idle connection cleanup
        - Thread-safe concurrent access
        - Automatic connection health checking
    """
    
    def __init__(self, addrs: List[str], max_conns: int = 10,
                 connect_timeout: float = 5.0, idle_timeout: float = 60.0):
        """
        Creates a new connection pool for the specified servers.
        
        Args:
            addrs: List of server addresses in "host:port" format
            max_conns: Maximum connections to maintain per server
            connect_timeout: Timeout for establishing new connections
            idle_timeout: How long connections can be idle before cleanup
        """
        self.addrs = addrs
        self.max_conns = max_conns
        self.connect_timeout = connect_timeout
        self.idle_timeout = idle_timeout
        self.pools: Dict[str, List[Connection]] = {}
        self.lock = threading.RLock()
        self.closed = False
        
        # Initialize empty pools for each server
        for addr in addrs:
            self.pools[addr] = []
    
    def get(self, addr: Optional[str] = None) -> Connection:
        """
        Retrieves a connection from the pool or creates a new one.
        
        It prefers reusing existing idle connections but will create new ones if needed.
        Stale connections are automatically discarded.
        
        Args:
            addr: Specific server address, or None to use the first available server
            
        Returns:
            Ready-to-use connection
            
        Raises:
            ClientClosedError: If pool is closed
            NetworkError: If connection cannot be established
        """
        with self.lock:
            if self.closed:
                raise ClientClosedError()
            
            # If no specific address requested, use the first server
            if addr is None:
                if not self.addrs:
                    raise NetworkError("connect", "", Exception("No addresses available"))
                addr = self.addrs[0]
            
            # Ensure pool exists for this address
            if addr not in self.pools:
                self.pools[addr] = []
            
            pool = self.pools[addr]
            
            # Try to reuse an existing connection (LIFO order)
            while pool:
                conn = pool.pop()
                if conn.is_alive():
                    return conn
                conn.close()
            
            # No reusable connection available; create a new one
            return self._create_connection(addr)
    
    def _create_connection(self, addr: str) -> Connection:
        """
        Creates a new TCP connection to a server.
        
        Args:
            addr: Server address in "host:port" format
            
        Returns:
            New connection
            
        Raises:
            NetworkError: If connection fails
        """
        try:
            host, port_str = addr.rsplit(':', 1)
            port = int(port_str)
            
            sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
            sock.settimeout(self.connect_timeout)
            sock.connect((host, port))
            sock.settimeout(None)  # Reset to blocking mode
            
            return Connection(sock, addr)
        except socket.timeout as e:
            raise ConnectionTimeoutError(addr)
        except Exception as e:
            raise NetworkError("connect", addr, e)
    
    def put(self, conn: Optional[Connection]) -> None:
        """
        Returns a connection to the pool for reuse.
        
        The connection is only kept if:
            - The pool is not closed
            - The pool is not full
            - The connection hasn't been idle too long
        
        Otherwise, the connection is closed.
        
        Args:
            conn: Connection to return (None is safe)
        """
        if conn is None:
            return
        
        with self.lock:
            if self.closed:
                conn.close()
                return
            
            addr = conn.get_addr()
            if addr not in self.pools:
                conn.close()
                return
            
            pool = self.pools[addr]
            
            # Discard connection if pool is at capacity
            if len(pool) >= self.max_conns:
                conn.close()
                return
            
            # Discard connection if it's been idle too long
            if time.time() - conn.get_last_used() > self.idle_timeout:
                conn.close()
                return
            
            # Connection is healthy and pool has space; add it back
            pool.append(conn)
            
            # Trigger periodic cleanup
            self._clean_pool(addr)
    
    def _clean_pool(self, addr: str) -> None:
        """
        Removes stale and dead connections from a server pool.
        
        Args:
            addr: Server address
        """
        if addr not in self.pools:
            return
        
        pool = self.pools[addr]
        now = time.time()
        valid_conns = []
        
        for conn in pool:
            if now - conn.get_last_used() > self.idle_timeout or not conn.is_alive():
                conn.close()
            else:
                valid_conns.append(conn)
        
        self.pools[addr] = valid_conns
    
    def add_addr(self, addr: str) -> None:
        """
        Dynamically adds a new server address to the pool.
        
        This is useful for adding storage servers discovered at runtime.
        If the address already exists, this is a no-op.
        
        Args:
            addr: Server address in "host:port" format
        """
        with self.lock:
            if self.closed:
                return
            
            if addr in self.pools:
                return
            
            self.addrs.append(addr)
            self.pools[addr] = []
    
    def close(self) -> None:
        """
        Shuts down the connection pool and closes all connections.
        
        After close is called, get will raise ClientClosedError.
        It's safe to call close multiple times.
        """
        with self.lock:
            if self.closed:
                return
            
            self.closed = True
            
            for pool in self.pools.values():
                for conn in pool:
                    conn.close()
                pool.clear()

File: typescript_client/src/connection.ts
/**
 * FastDFS Connection Management
 * 
 * This module handles TCP connections to FastDFS servers with connection pooling,
 * automatic reconnection, and health checking.
 */

import * as net from 'net';
import { NetworkError, ConnectionTimeoutError, ClientClosedError } from './errors';

/**
 * Represents a TCP connection to a FastDFS server (tracker or storage)
 * 
 * It wraps a net.Socket with additional metadata and thread-safe operations.
 * Each connection tracks its last usage time for idle timeout management.
 */
export class Connection {
  private socket: net.Socket;
  private addr: string;
  private lastUsed: number;
  private closed: boolean = false;

  constructor(socket: net.Socket, addr: string) {
    this.socket = socket;
    this.addr = addr;
    this.lastUsed = Date.now();
  }

  /**
   * Transmits data to the server with optional timeout
   * 
   * This method updates the lastUsed timestamp.
   */
  async send(data: Buffer, timeout: number = 30000): Promise<void> {
    if (this.closed) {
      throw new NetworkError('write', this.addr, new Error('Connection closed'));
    }

    return new Promise((resolve, reject) => {
      const timer = setTimeout(() => {
        reject(new NetworkError('write', this.addr, new Error('Write timeout')));
      }, timeout);

      this.socket.write(data, (err) => {
        clearTimeout(timer);
        if (err) {
          reject(new NetworkError('write', this.addr, err));
        } else {
          this.lastUsed = Date.now();
          resolve();
        }
      });
    });
  }

  /**
   * Reads up to 'size' bytes from the server
   * 
   * This method may return fewer bytes than requested.
   * Use receiveFull if you need exactly 'size' bytes.
   */
  async receive(size: number, timeout: number = 30000): Promise<Buffer> {
    if (this.closed) {
      throw new NetworkError('read', this.addr, new Error('Connection closed'));
    }

    return new Promise((resolve, reject) => {
      const timer = setTimeout(() => {
        reject(new NetworkError('read', this.addr, new Error('Read timeout')));
      }, timeout);

      const onData = (data: Buffer) => {
        clearTimeout(timer);
        this.socket.removeListener('data', onData);
        this.socket.removeListener('error', onError);
        this.lastUsed = Date.now();
        resolve(data.slice(0, size));
      };

      const onError = (err: Error) => {
        clearTimeout(timer);
        this.socket.removeListener('data', onData);
        this.socket.removeListener('error', onError);
        reject(new NetworkError('read', this.addr, err));
      };

      this.socket.once('data', onData);
      this.socket.once('error', onError);
    });
  }

  /**
   * Reads exactly 'size' bytes from the server
   * 
   * This method blocks until all bytes are received or an error occurs.
   * The timeout applies to the entire operation, not individual reads.
   */
  async receiveFull(size: number, timeout: number = 30000): Promise<Buffer> {
    if (this.closed) {
      throw new NetworkError('read', this.addr, new Error('Connection closed'));
    }

    return new Promise((resolve, reject) => {
      const chunks: Buffer[] = [];
      let totalReceived = 0;

      const timer = setTimeout(() => {
        this.socket.removeListener('data', onData);
        this.socket.removeListener('error', onError);
        reject(new NetworkError('read', this.addr, new Error('Read timeout')));
      }, timeout);

      const onData = (data: Buffer) => {
        chunks.push(data);
        totalReceived += data.length;

        if (totalReceived >= size) {
          clearTimeout(timer);
          this.socket.removeListener('data', onData);
          this.socket.removeListener('error', onError);
          this.lastUsed = Date.now();
          
          const result = Buffer.concat(chunks);
          resolve(result.slice(0, size));
        }
      };

      const onError = (err: Error) => {
        clearTimeout(timer);
        this.socket.removeListener('data', onData);
        this.socket.removeListener('error', onError);
        reject(new NetworkError('read', this.addr, err));
      };

      this.socket.on('data', onData);
      this.socket.once('error', onError);
    });
  }

  /**
   * Terminates the connection and releases resources
   * 
   * It's safe to call close multiple times.
   */
  close(): void {
    if (!this.closed) {
      this.closed = true;
      this.socket.destroy();
    }
  }

  /**
   * Performs a check to determine if the connection is still valid
   */
  isAlive(): boolean {
    return !this.closed && !this.socket.destroyed && this.socket.writable && this.socket.readable;
  }

  /**
   * Returns the timestamp of the last send or receive operation
   */
  getLastUsed(): number {
    return this.lastUsed;
  }

  /**
   * Returns the server address this connection is connected to
   */
  getAddr(): string {
    return this.addr;
  }
}

/**
 * Manages a pool of reusable connections to multiple servers
 * 
 * It maintains separate pools for each server address and handles:
 *   - Connection reuse to minimize overhead
 *   - Idle connection cleanup
 *   - Thread-safe concurrent access
 *   - Automatic connection health checking
 */
export class ConnectionPool {
  private addrs: string[];
  private maxConns: number;
  private connectTimeout: number;
  private idleTimeout: number;
  private pools: Map<string, Connection[]>;
  private closed: boolean = false;

  constructor(
    addrs: string[],
    maxConns: number = 10,
    connectTimeout: number = 5000,
    idleTimeout: number = 60000
  ) {
    this.addrs = addrs;
    this.maxConns = maxConns;
    this.connectTimeout = connectTimeout;
    this.idleTimeout = idleTimeout;
    this.pools = new Map();

    // Initialize empty pools for each server
    for (const addr of addrs) {
      this.pools.set(addr, []);
    }
  }

  /**
   * Retrieves a connection from the pool or creates a new one
   * 
   * It prefers reusing existing idle connections but will create new ones if needed.
   * Stale connections are automatically discarded.
   */
  async get(addr?: string): Promise<Connection> {
    if (this.closed) {
      throw new ClientClosedError();
    }

    // If no specific address requested, use the first server
    if (!addr) {
      if (this.addrs.length === 0) {
        throw new NetworkError('connect', '', new Error('No addresses available'));
      }
      addr = this.addrs[0];
    }

    // Ensure pool exists for this address
    if (!this.pools.has(addr)) {
      this.pools.set(addr, []);
    }

    const pool = this.pools.get(addr)!;

    // Try to reuse an existing connection (LIFO order)
    while (pool.length > 0) {
      const conn = pool.pop()!;
      if (conn.isAlive()) {
        return conn;
      }
      conn.close();
    }

    // No reusable connection available; create a new one
    return this.createConnection(addr);
  }

  /**
   * Creates a new TCP connection to a server
   */
  private async createConnection(addr: string): Promise<Connection> {
    const [host, portStr] = addr.split(':');
    const port = parseInt(portStr, 10);

    return new Promise((resolve, reject) => {
      const socket = new net.Socket();
      const timer = setTimeout(() => {
        socket.destroy();
        reject(new ConnectionTimeoutError(addr));
      }, this.connectTimeout);

      socket.connect(port, host, () => {
        clearTimeout(timer);
        socket.setNoDelay(true);
        socket.setKeepAlive(true);
        resolve(new Connection(socket, addr));
      });

      socket.on('error', (err) => {
        clearTimeout(timer);
        reject(new NetworkError('connect', addr, err));
      });
    });
  }

  /**
   * Returns a connection to the pool for reuse
   * 
   * The connection is only kept if:
   *   - The pool is not closed
   *   - The pool is not full
   *   - The connection hasn't been idle too long
   * 
   * Otherwise, the connection is closed.
   */
  put(conn: Connection | null): void {
    if (!conn) {
      return;
    }

    if (this.closed) {
      conn.close();
      return;
    }

    const addr = conn.getAddr();
    const pool = this.pools.get(addr);

    if (!pool) {
      conn.close();
      return;
    }

    // Discard connection if pool is at capacity
    if (pool.length >= this.maxConns) {
      conn.close();
      return;
    }

    // Discard connection if it's been idle too long
    if (Date.now() - conn.getLastUsed() > this.idleTimeout) {
      conn.close();
      return;
    }

    // Connection is healthy and pool has space; add it back
    pool.push(conn);

    // Trigger periodic cleanup
    this.cleanPool(addr);
  }

  /**
   * Removes stale and dead connections from a server pool
   */
  private cleanPool(addr: string): void {
    const pool = this.pools.get(addr);
    if (!pool) return;

    const now = Date.now();
    const validConns: Connection[] = [];

    for (const conn of pool) {
      if (now - conn.getLastUsed() > this.idleTimeout || !conn.isAlive()) {
        conn.close();
      } else {
        validConns.push(conn);
      }
    }

    this.pools.set(addr, validConns);
  }

  /**
   * Dynamically adds a new server address to the pool
   * 
   * This is useful for adding storage servers discovered at runtime.
   * If the address already exists, this is a no-op.
   */
  addAddr(addr: string): void {
    if (this.closed) {
      return;
    }

    if (this.pools.has(addr)) {
      return;
    }

    this.addrs.push(addr);
    this.pools.set(addr, []);
  }

  /**
   * Shuts down the connection pool and closes all connections
   * 
   * After close is called, get will throw ClientClosedError.
   * It's safe to call close multiple times.
   */
  close(): void {
    if (this.closed) {
      return;
    }

    this.closed = true;

    for (const pool of this.pools.values()) {
      for (const conn of pool) {
        conn.close();
      }
      pool.length = 0;
    }
  }
}